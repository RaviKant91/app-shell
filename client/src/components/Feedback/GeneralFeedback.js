import React from "react";
import Button from "@material-ui/core/Button";
import DialogTitle from "@material-ui/core/DialogTitle";
import DialogContent from "@material-ui/core/DialogContent";
import DialogActions from "@material-ui/core/DialogActions";
import Dialog from "@material-ui/core/Dialog";
import withStyles  from '@material-ui/core/styles/withStyles';
import TextField from '@material-ui/core/TextField';
import FormControl from '@material-ui/core/FormControl';
import IconButton from '@material-ui/core/IconButton';
import CloseIcon from '@material-ui/icons/Close';
import ThankyouFeedback from "./ThankyouFeedback";
import SessionTimeout from "./SessionTimeout";
import "./FeedbackSubPopup.css";
import { withTranslation } from 'react-i18next';

const styles = {
  GeneralFeedbackLogTitle: {
    textAlign: "center !important",
    paddingTop:"0px"
  },
  GeneralFeedbackLogTitleText:{
    fontFamily:"var(--font-family-subtitle-medium)",
    fontStyle: "normal",
    fontWeight: "normal",
    fontSize: "16px",
    lineHeight: "24px",
    alignItems: "center",
    letterSpacing: "0.1px",
    color: "var(--color-text-common-primary)",
    paddingTop:"0px"
  },
  GeneralFeedbackDialog: {
    margin: "15px"
  },
  GeneralFeedbackDialogcontent: {
    width: "400px !important",
    padding:"0px 24px !important"
  },
  root: {
    minWidth: "275px !important",
  },
  formControl: {
    minWidth: "120 px",
    width: "100% !important",
    marginLeft: "0 px",
    marginBottom : "15px",
    "& .MuiFilledInput-multiline":{
      border: "1px solid var(--color-border-form-default)",
      "&:hover":{
        border: "1px solid var(--color-border-form-hover) !important"
      }
    }
  },
  selectEmpty: {
    marginTop: "15px"
  },
  FeedbackButton: {
    backgroundColor: 'var(--color-fill-cta-primary-default)', 
    borderRadius: "4px",
    color: 'var(--color-text-common-inverse-primary)',
    fontFamily:"Poppins",
    fontStyle: "normal",
    fontWeight: "500",
    fontSize: "12px",
    lineHeight: "18px",   
    letterSpacing:"0.1px",
    height:"36px",
    position: 'absolute',
    left: "62% !important",
    textTransform: "uppercase",
    '&:hover': {backgroundColor: "var(--color-fill-cta-primary-hover)",
      borderColor: "var(--color-border-cta-primary-hover)"}
  },
  media: {
    height: 140,
  },
  closeButton: {
    textAlign: "right",
    "& .MuiIconButton-root":{
      color: 'var(--color-text-common-primary)'
    }
   },
   FeedbackButtonAction:{
     paddingBottom:"30px",
   },
   paper: {
    width: "40% !important",
    maxHeight: 435,
  },
  formError:{
    color:"#E12D39", margin: "5px", fontFamily:"Poppins", fontWeight: 400,fontSize: "0.75rem",lineHeight: 1.66,letterSpacing: "0.03333em", textAlign: "left", marginTop: "3px",marginRight: 0, marginBottom: 0, marginLeft: 0
  },
  feedbackCharCount:{
    float:"right", margin: "5px",fontFamily:"Poppins", fontWeight: 400, fontSize: "0.80rem", lineHeight: 1.66, letterSpacing: "0.03333em",marginTop: "3px", marginRight: 0, marginBottom: 0,marginLeft: 0, color: "var(--color-text-common-secondary)"
    },
  CommentformError:{
    float:"left", color:"#E12D39", margin: "5px", fontFamily:"Poppins", fontWeight: 400,fontSize: "0.75rem",lineHeight: 1.66,letterSpacing: "0.03333em",textAlign: "left", marginTop: "3px", marginRight: 0, marginBottom: 0, marginLeft: 0 
  }
};

const getUserInfo = ()=>{
 const getUserData = localStorage.getItem("UserInfo");
 if(getUserData){
   return JSON.parse(getUserData);
 }else{
   return[];
 }
}


function GeneralFeedback(props) {
  
  const { 
    classes,
    onClose,
    open,
    t,
    getBrowserInfo,
    ...other
     } = props;

  const [userInfo, setUserInfo] = React.useState(getUserInfo());
  const [userId, setUserId] = React.useState(`${userInfo.name}`);
  const [ErrorFlag, setErrorFlag] = React.useState(false);
  const [userFullname, setUserFullname] = React.useState();
  const [thankuFeedbackopen, setThankuFeedbackopen] = React.useState(false);
  const initialValues = { getUserName: userFullname, getComment: "" };
  const [formValues, setFormValues] = React.useState(initialValues);
  const [formErrors, setFormErrors] = React.useState({});
  const [isSubmit, setIsSubmit] = React.useState(false);
  const [isCharCount, setCharCount] = React.useState(0);
  const [dialogShow, setdialogShow] = React.useState(false);
  const [eventType, seteventType] = React.useState("");
  
  React.useEffect(() => {
    if (userInfo.firstName === "" && userInfo.lastName === "" ) {
          setUserFullname(userId);
        } else{
          setUserFullname(`${userInfo.firstName} ${userInfo.lastName}`);
        }
  }, []);
  const formHandleChange = (e) => {
    const { name, value } = e.target;
    setFormValues({ ...formValues, [name]: value });
  };
  const formHandleSubmit = (e) => {
    e.preventDefault();
    if(formValues.getUserName !== "" && formValues.getComment !== ""){
     fetch(`${window.location.pathname}Service-hub/feedback-proxy/api/feedbacks`,{
        method: 'POST',
        headers: {
          'Accept': 'application/json',
          'Content-Type': 'application/json',
          'Cache-Control':' no-cache',
          'strapi':'true'
        },
        body: JSON.stringify({data:{user_info: {
          userName : (userInfo.firstName !== "" && userInfo.lastName !== "")  ? `${userInfo.firstName} ${userInfo.lastName}` : `${userInfo.name}`,
          email:userInfo.email
        }, browser_info : getBrowserInfo, user_comment: formValues.getComment, microapp_info: window.currentAppObject,feedback_type:'General Feedback'
      }})
    })
    .then(response =>response.json())
    .then(data => {
      /* istanbul ignore if */
      if(data.timeoutmsg === "success timeout"){
        setdialogShow(true);
        seteventType("session-timeout");
      }
      else if(data.data!==null) {
        setIsSubmit(true);
        setThankuFeedbackopen(true);
        onClose();
        setFormValues(initialValues);
        setCharCount(0);
        setFormErrors({});
      } 
      else{
        /* istanbul ignore if */
        if(data.error){
          setErrorFlag(true);
          setIsSubmit(true);
          onClose();
          setThankuFeedbackopen(true);
          setFormValues(initialValues);
          setCharCount(0);
          setFormErrors({});
        }
      }
       
    }).catch((error)=> {
        setErrorFlag(true);
        setIsSubmit(true);
        onClose();
        setThankuFeedbackopen(true);
        setFormValues(initialValues);
        setCharCount(0);
        setFormErrors({});
    });     
    }else{
      setFormErrors(validate(formValues));
    }
  };

  React.useEffect(() => {
    /*istanbul ignore if*/
    if (Object.keys(formErrors).length === 0 && isSubmit) {
      
    }
  }, [formErrors]);
  
 const validate = (values) => {
    const errors = {};
    /* istanbul ignore else */
if (!values.getComment) {
      errors.getComment = t("ErrorMasgFeedbackComment");
    } 
   return errors;
  };

  const closeThankuFeedback = () => {
    setThankuFeedbackopen(false);
    setErrorFlag(false);
  };

  const handleCancel = () => {
    onClose();
    setFormValues(initialValues);
    setIsSubmit(false);
    setFormErrors({});
    setCharCount(0);
  };

  let getCommentGenHandle = (e)=>{
    setCharCount(e.target.value.length);
    formHandleChange(e);
  }
  /* istanbul ignore next */
  const setdialogShowPopup = () => {
    setdialogShow(false);
  };

  let GeneralCommentformError =   isCharCount === 0 ? <p id="generalCommentFormError" className={classes.CommentformError}>{formErrors.getComment}</p> : null;
  /* istanbul ignore if */
  if(eventType=="session-timeout"){
    return(
        <SessionTimeout
         dialogShow={dialogShow}
         setdialogShow={setdialogShowPopup}
        />
    )
 }
  return (
  <>
   <form>
    <Dialog open={open} {...other} className={classes.GeneralFeedbackDialog}>
      <div className={classes.closeButton}>
        <IconButton aria-label="close"
         onClick={handleCancel}>
            <CloseIcon />
        </IconButton>
      </div>
      <DialogTitle className={classes.GeneralFeedbackLogTitle}>
        <h2 id="generalFeedbackTitle" className={classes.GeneralFeedbackLogTitleText}> {t("ShareYourFeedback")}</h2>
      </DialogTitle>
 
      <DialogContent className={classes.GeneralFeedbackDialogcontent}>
      <FormControl className={classes.formControl} id="GeneralFeedbackUsername">
      <TextField
          disabled
          defaultValue={userFullname} 
          size="small"
          id="emailAddressGeneralFeedback"
          name="getUserName"
          value={formValues.getUserName}
          onChange={formHandleChange}
          InputProps={{ disableUnderline: true }}
      />
     </FormControl>

      <FormControl className={classes.formControl}>
        <TextField
            placeholder={t("LeaveUsYourComment")}
            multiline 
            variant="filled"  
            rows={4}
            id="leaveCommentGeneralFeedback"
            name="getComment"
            value={formValues.getComment}
            onChange={getCommentGenHandle}
            inputProps={{
              maxLength: 1024
            }}
        />
        <div>
           <p className={classes.feedbackCharCount}>{isCharCount}/1024</p> 
           {GeneralCommentformError}
        </div>
      </FormControl>
      </DialogContent>
      <DialogActions className={classes.FeedbackButtonAction}>
          <Button
            variant="contained"
            className={classes.FeedbackButton}
            onClick={formHandleSubmit}
            color="primary"
            type="submit"
          >
                {t("SendFeedback")}
          </Button>
      </DialogActions>
  </Dialog>
 </form>
  <ThankyouFeedback
      classes={{
        paper: classes.paper
      }}
      openThanku={thankuFeedbackopen}
      errorFlag={ErrorFlag}
      onCloseThanku={closeThankuFeedback}
  />
</>
  );
}

export default withTranslation()(withStyles(styles)(GeneralFeedback));
