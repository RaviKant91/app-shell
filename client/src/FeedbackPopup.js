import React, { useRef } from "react";
import { withStyles } from "@material-ui/core/styles";
import Popover, { PopoverAnimationVertical } from "material-ui/Popover";
import { withTranslation } from 'react-i18next';
import IconButton from '@material-ui/core/IconButton';
import CloseIcon from '@material-ui/icons/Close';
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemAvatar from '@material-ui/core/ListItemAvatar';
import ListItemText from '@material-ui/core/ListItemText';
import { blue } from '@material-ui/core/colors';
import GeneralFeedback from "./components/Feedback/GeneralFeedback";
import RateReviewOutlinedIcon from '@material-ui/icons/RateReviewOutlined';
import generalIcon from './assets/General_icon.svg';
import drawIcon from './assets/Draw_icon.svg';
import CardMedia from '@material-ui/core/CardMedia';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';


const styles = {
    feedbackPopup: {
      position: "relative !important",
      float: "right",
      margin: "0 3em",
      width: '385px',
      height: '355px',
      top: "74px !important",
      boxShadow: "var(--effect-drop-shadow-elevation-extra-high) !important",
      borderRadius: "4px !important",
      overflowX: "hidden !important",
      padding: '10px 20px',
      backgroundColor: 'var(--color-fill-common-secondary) !important'
    },
    closeButton: {
        textAlign: "right",
        marginRight: "-10px",
        "& .MuiIconButton-root":{
          color: 'var(--color-text-common-primary)'
        }
    },
    root: {
        minWidth: 275,
    },
    bullet: {
        display: 'inline-block',
        margin: '0 2px',
        transform: 'scale(0.8)',
    },
    title: {
        fontSize: 14,
    },
    pos: {
        marginBottom: 12,
    },
    avatar: {
        backgroundColor: blue[100],
        color: blue[600],
    },
    CompleteSnapshotPopup: {
        width: "40% !important",
        maxHeight: 435
    },
    feednacktitle:{
      textAlign:"center",
      paddingTop: "0px",
      paddingBottom: "43px",
      color: 'var(--color-text-common-primary)'
    },
    feedbackicon:{
      width: "32px",
      height: "32px",
      marginBottom: "-11px",
      marginRight: "5px"
    },
    feedbacktext:{
      display: "inlineBlock",
      fontFamily: "Poppins",
      fontStyle: "normal",
      fontWeight: "600",
      fontSize: "20px",
      lineHeight: "24px",
      letterSpacing: "-0.5px",
      color: "var(--color-text-common-primary)"
    }

};
const useStyles = makeStyles((theme) => ({
  root: {
    width: "372px",
    height: "103px",
    left: "2px",
    top: "101px",
    boxSizing: "border-box"
  },
  paper: {
    width: "40% !important",
    maxHeight: 435
  },
  FeedbackDropIcon: {
    width: "32px",
    height: "32px",
  },
  FeedbackDropsubtitle: {
    marginBottom: 12,
    fontSize: 13,
    fontFamily:"Noto Sans",
    fontStyle: "normal",
    fontWeight: "normal",
    lineHeight: "20px",
    color: "var(--color-text-common-primary)"
  },
  avatar: {
    backgroundColor: blue[100],
    color: blue[600],
  },
  CompleteSnapshotPopupContainer:{

  },
  FeedbackDropmaintitle: {
    fontFamily: "Poppins",
    fontStyle: "normal",
    fontWeight: "600",
    fontSize: "16px",
    lineHeight: "22px",
    letterSpacing: "-0.25px",
    color: "var(--color-text-common-primary)"
  },
  FeedbackDropContent:{
    padding: 10,
    textAlign:"left"
  },
  GenFeedbackDropContent:{
    padding: 10,
    paddingRight: 0,
    textAlign:"left"
  },
  FeedbackDropListItem:{
    paddingTop:0,
    backgroundColor: 'var(--color-fill-common-secondary) !important'
  },
  FeedbackDropDivIcon:{
    paddingBottom: 20,
  }
}));

const FeedbackPopup = (props) => {
    const {
      classes,
      isOpen,
      setFeedbackPopupOpen,
      closeFeedback,
      currentAppObject,
      browserInfo,
      t
    } = props;

  const myclasses = useStyles();
  
  const [GeneralFeedbackopen, setGeneralFeedbackopen] = React.useState(false);

  const bhScreenCapture = () => {
    document.getElementById('completesnapshotDombtn').click();
    document.getElementById('captureButton').click();
    setFeedbackPopupOpen(false);
  };

  const closeGeneralFeedback = () => {
    setGeneralFeedbackopen(false);
  };
  const handleClickGeneralFeedback = () => {
    setFeedbackPopupOpen(false);
    setGeneralFeedbackopen(true);
  };

 

    return (
     <MuiThemeProvider>
     {
      isOpen &&
     <Popover
        open={isOpen}
        anchorOrigin={{ horizontal: "left", vertical: "bottom" }}
        transformOrigin={{
          vertical: "top",
          horizontal: "right",
        }}
        className={classes.feedbackPopup}
        onRequestClose={() => {
            setFeedbackPopupOpen(false);
        }}
        animation={PopoverAnimationVertical}
      >
      <div id="feedbackPopupWrapper">
      <div className={classes.closeButton}>
        <IconButton aria-label="close"
          onClick={() => {
              closeFeedback();
          }}>
            <CloseIcon />
        </IconButton>
      </div>


    <CardContent id="feedbackHeaderTitle" className={classes.feednacktitle}>
      
      <span><RateReviewOutlinedIcon className={classes.feedbackicon}/></span > 
      <span className={classes.feedbacktext}>{t("Feedback")}</span> 

    </CardContent>

    <div className={myclasses.root}>
      <List>
        <ListItem
          id="specificFeedback"
          button
          aria-haspopup="true"
          onClick={bhScreenCapture}
          className={myclasses.FeedbackDropListItem}
          style={{backgroundColor:"#ffffff"}}
        >
             <ListItemAvatar className={myclasses.FeedbackDropDivIcon}>
                <CardMedia
                    className={myclasses.FeedbackDropIcon}
                    image={drawIcon}
                />
            </ListItemAvatar>
            <ListItemText>
              < CardContent className={myclasses.FeedbackDropContent}>
                <Typography id="specificFeedbackTitle" className={myclasses.FeedbackDropmaintitle} gutterBottom>
                {t("SpecificFeedback")}
                </Typography>
                <Typography id="specificFeedbackDescription" className={myclasses.FeedbackDropsubtitle} color="textSecondary">
                {t("SpecificFeedbackSubTitle")}
                </Typography>
              </CardContent>
            </ListItemText>
        </ListItem>
      </List>
    </div>
    
    <div className={myclasses.root}>
      <List>
        <ListItem
          id="generalFeedback"
          button
          aria-haspopup="true"
          onClick={handleClickGeneralFeedback}
          className={myclasses.FeedbackDropListItem}
          style={{backgroundColor:"#ffffff"}}
        >
            <ListItemAvatar className={myclasses.FeedbackDropDivIcon}>
                
            <CardMedia
                    className={myclasses.FeedbackDropIcon}
                    image={generalIcon}
              />
               
            </ListItemAvatar>
            <ListItemText>
              < CardContent className={myclasses.GenFeedbackDropContent}>
                <Typography  id="generalFeedbackTitle" className={myclasses.FeedbackDropmaintitle} gutterBottom>
                {t("GeneralFeedback")}
                </Typography>
                <Typography  id="generalFeedbackDescription" className={myclasses.FeedbackDropsubtitle} color="textSecondary">
                {t("GeneralFeedbackSubTitle")}
                </Typography>
              </CardContent>
            </ListItemText>
        </ListItem>
      </List>
    </div>
    </div>
    </Popover>
    }
    <GeneralFeedback
          classes={{
            paper: myclasses.paper
          }}
          open={GeneralFeedbackopen}
          getBrowserInfo= {browserInfo}
          onClose={closeGeneralFeedback}
      />
    </MuiThemeProvider>
    );
  };

  export default withTranslation()(withStyles(styles)(FeedbackPopup));
