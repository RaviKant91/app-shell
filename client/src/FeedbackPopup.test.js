import React from "react";
import _ from "lodash";
import { mount, shallow } from 'enzyme';
import { withStyles } from "@material-ui/core/styles";
import Popover, { PopoverAnimationVertical } from "material-ui/Popover";
import { withTranslation } from 'react-i18next';
import IconButton from '@material-ui/core/IconButton';
import CloseIcon from '@material-ui/icons/Close';
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemAvatar from '@material-ui/core/ListItemAvatar';
import ListItemText from '@material-ui/core/ListItemText';
import { blue } from '@material-ui/core/colors';
import GeneralFeedback from "./components/Feedback/GeneralFeedback";
import RateReviewOutlinedIcon from '@material-ui/icons/RateReviewOutlined';
import generalIcon from './assets/General_icon.svg';
import drawIcon from './assets/Draw_icon.svg';
import CardMedia from '@material-ui/core/CardMedia';
import FeedbackPopup from './FeedbackPopup'

jest.mock("react-i18next", () => ({
    withTranslation: () => (Component) => {
        Component.defaultProps = { ...Component.defaultProps, t: (key) => key };
        return Component;
    },
}));

describe('Feedback Popup Start', () => {
    const CommonComponentProps = {
        isOpen: true,
        setFeedbackPopupOpen: jest.fn(),
        closeFeedback: jest.fn(),
        t: jest.fn(),
        browserInfo: {
            "Browsername": "Chrome",
            "version": "108.0.0.0",
            "majorVersion": 108
        },
        classes: {
            "feedbackPopup": "jss83",
            "closeButton": "jss84",
            "root": "jss85",
            "bullet": "jss86",
            "title": "jss87",
            "pos": "jss88",
            "avatar": "jss89",
            "CompleteSnapshotPopup": "jss90",
            "feednacktitle": "jss91",
            "feedbackicon": "jss92",
            "feedbacktext": "jss93"
        },
        tReady: true
    };

    it('verify for FeedbackPopup when loads', () => {
        const component = mount(<FeedbackPopup {...CommonComponentProps} />);
        expect(component).toMatchSnapshot();
    });

    
});