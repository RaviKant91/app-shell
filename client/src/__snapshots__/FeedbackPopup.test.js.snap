// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feedback Popup Start verify for FeedbackPopup when loads 1`] = `ReactWrapper {}`;
